Review and Merge how-to - Custom Save Button (#5422)
mendix/docs: content/en/docs/refguide/modeling/pages/button-widgets/create-a-custom-save-button.md

---
title: "Creating a Custom Save Button with a Microflow"

> url: /refguide/creating-a-custom-save-button/

Reviewer: The title does not fit in the sidebar, please add a "linktitle" to metadata that is shorter. For example, Creating a Custom Save Button.

Author: Advice taken :)

description: "Describes how to create a custom save button by utilizing microflows."
tags: ["microflow", "logic", "save button"]
aliases:
    - /howto/logic-business-rules/create-a-custom-save-button
    - /howto/logic-business-rules/create-a-custom-save-button.html
---
## 1 Introduction

Mendix uses visual models called microflows to define the logic of your application. A [microflow](/refguide/microflows/) is a visual way of expressing what traditionally would be written in code. 

> This document explains how you can create a custom Save button on a detail page using a microflow. You will replace the default Save button with a Save button that shows a custom message to end-users after they click it. 

Reviewer: 1. **Save** should be in bold as a specific UI element.
2. The Future tense is better to be replaced by the Present tense according to our style guide. Also "you will replace" sounds more as a how-to style or even Academy.

Author: 1. **Save** is bolded now for all buttons. 
2. I changed "you will replace" to "you need to replace". I think the latter is less of a how-to style. Please let me know if you don't agree.

> This document explains how you can create a custom Save button on a detail page using a microflow. You will replace the default Save button with a Save button that shows a custom message to end-users after they click it. 

Reviewer: ```suggestion
This document explains how you can create a custom Save button on a detail page using a microflow. You will replace the default Save button with a Save button that shows a custom message to end-users after they click the button. 
```

> Before you continue, make sure you have completed the following:

Reviewer: These two documents are quite long. Are they necessary to be completed from top to bottom? I suggest adding two points:
* Create a Customer entity in your domain model (for more information, see....)
Optionally a screenshot of the entity
* Create a detail page for the Customer entity (for more informatuon, see...)
Optionally a screenshot with a page example

Author: Thanks for this comment. I've addressed this comment in this commit: https://github.com/mendix/docs/pull/5422/commits/ecf893acf158792b1d8389b641012674574d9869 (line 19-25)

Reviewer: Thanks! It looks more like a bulleted list to me, numbered one are steps usually, unless this is important they do the first one first.

Author: As you mentioned above, the order actually does matter. You need to create the entity first before creating a detail page for it. I'll keep the numbered list there for this reason.

* Add a **Customer** entity to your domain model (for guidance, see [Create a basic data layer](/howto/data-models/create-a-basic-data-layer/))
* [Create overview and detail pages for the Customer entity](/howto/front-end/create-your-first-two-overview-and-detail-pages/)

## 2 Replacing the Default Save Button with a Custom One

> To replace the default **Save** button with a custom one, follow these steps:

Reviewer: I would add a bit more explanation of what they are going to do. That a detail page has a default **Save** button, but they need to replace it, etc.

Author: Thanks for the advice. I've added a brief explanation there too.

1. Open the **Customer_Detail** page:

    {{< figure src="/attachments/refguide/modeling/pages/button-widgets/create-a-custom-save-button/customer-detail.png" width="400px" >}}

2. Right-click the **Save** button and select **Delete** to remove it.

> 3. Right-click the drop-zone under the **Cancel** button.

Reviewer: In other places of the refguide and how-tos we seem to navigate through **Toolbox** mostly rather than working area (canvas). Affects points 3 and 4 here.

Author: I changed the steps to navigate through **Toolbox** now.

4. Select **Add widget** and then select **Call microflow button**.
5. In the **Select Microflow** dialog box, click **New** to create a new microflow.
6. Set the **Name** as *Customer_Save*.

    {{< figure src="/attachments/refguide/modeling/pages/button-widgets/create-a-custom-save-button/customer-save.png" width="400px" >}}

7. After the button is created, open its **Properties** pane.
8. Change the **Caption** property to *Save*:

    {{< figure src="/attachments/refguide/modeling/pages/button-widgets/create-a-custom-save-button/button-properties.png" >}}

9. Set a floppy disk as the **Icon**:

    {{< figure src="/attachments/refguide/modeling/pages/button-widgets/create-a-custom-save-button/icon.png" >}}

You have created a button that calls a microflow.

## 3 Recreating the Default Save Behavior

> To recreate the default Save behavior, do the following:

Reviewer: Perhaps a bit more explanation, why they need to recreate a default behaviour first?

Author: Good idea! I've added a brief explanation there.

> 1. Right-click the new **Save** button and select **Go to on click microflow** to open the new microflow. The microflow should look like this:

Reviewer: ```suggestion
1. Right-click the new **Save** button and select **Go to on click microflow** to open the new microflow. An example of how the microflow should like is below:
```

    {{< figure src="/attachments/refguide/modeling/pages/button-widgets/create-a-custom-save-button/created-microflow.png" >}}

> 2. Open the **Toolbox** and select the **Commit object(s)** activity there.

Reviewer: ```suggestion
2. Open the **Toolbox** and search for the **Commit object(s)** activity there.
```

6. Set **Refresh in client** to **Yes** and click **OK**:

    {{< figure src="/attachments/refguide/modeling/pages/button-widgets/create-a-custom-save-button/commit-object-properties.png" width="400px" >}}

7. Open the **Toolbox**, and find the **Close page** activity there.
8. Drag the **Close page** activity into the flow after the **Commit object(s)** activity:

    {{< figure src="/attachments/refguide/modeling/pages/button-widgets/create-a-custom-save-button/close-page-activity.png" width="400px ">}}

You have recreated the default **Save** button logic.

## 4 Extending the Microflow with a Custom Message

To extend the logic with a custom message that is shown to end-users when they click the button, do the following:

1. Open **Toolbox**, find the **Show message** activity, and drag it after the **Close page** activity.
2. Double-click the **Show message** activity to open its properties.
3. In the **Template** column, type *Customer is saved!*.

    {{< figure src="/attachments/refguide/modeling/pages/button-widgets/create-a-custom-save-button/show-message-properties.png" width="400px" >}}

4. Click **OK** to save the changes. 

> Congratulations! You have customized the **Save** button using a microflow: 

Reviewer: Not sure if "congratulations" is also a how-to-specific format.

Author: I kept it in the doc. I think it is fine to leave it there for now.
